subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1250 @@
+			if (msg_rx->response_cb != NULL) {
+				msg_rx->response_cb(
+					(struct audio_module_handle_private *)msg_rx->tx_handle,
+					&msg_rx->audio_data);
+			}
+
+			LOG_ERR("Data process error in module %s, ret %d", handle->name, ret);
+			continue;
+		}
+
+		if (msg_rx->response_cb != NULL) {
+			msg_rx->response_cb((struct audio_module_handle_private *)msg_rx->tx_handle,
+					    &msg_rx->audio_data);
+		}
+
+		data_fifo_block_free(handle->thread.msg_rx, (void **)&msg_rx);
+	}
+
+	CODE_UNREACHABLE;
+}
+
+/**
+ * @brief The thread that processes inputs and outputs the data from the module.
+ *
+ * @param handle[in/out]  The handle for this modules instance.
+ *
+ * @return 0 if successful, error value.
+ */
+static void module_thread_in_out(struct audio_module_handle *handle, void *p2, void *p3)
+{
+	int ret;
+	struct audio_module_message *msg_rx;
+	struct audio_data audio_data;
+	void *data;
+	size_t size;
+
+	__ASSERT(handle != NULL, "Module task has NULL handle");
+	__ASSERT(handle->description->functions->data_process != NULL,
+		 "Module task has NULL process function pointer");
+
+	/* Execute thread. */
+	while (1) {
+		data = NULL;
+
+		LOG_DBG("Module %s is waiting for audio data", handle->name);
+
+		/* Get a new input message.
+		 * Since this input message is queued outside the module, this will then control the
+		 * data flow.
+		 */
+		ret = data_fifo_pointer_last_filled_get(handle->thread.msg_rx, (void **)&msg_rx,
+							&size, K_FOREVER);
+		__ASSERT(ret, "Module %s error in getting last filled", handle->name);
+
+		LOG_DBG("Module %s new audio data received", handle->name);
+
+		/* Get a new output buffer. */
+		ret = k_mem_slab_alloc(handle->thread.data_slab, (void **)&data, K_NO_WAIT);
+		__ASSERT(ret, "No free data buffer for module %s, dropping input, ret %d",
+			 handle->name, ret);
+
+		/* Configure new audio audio_data. */
+		audio_data.data = data;
+		audio_data.data_size = handle->thread.data_size;
+
+		/* Process the input audio data into the output audio data. */
+		ret = handle->description->functions->data_process(
+			(struct audio_module_handle_private *)handle, &msg_rx->audio_data,
+			&audio_data);
+		if (ret) {
+			if (msg_rx->response_cb != NULL) {
+				msg_rx->response_cb(
+					(struct audio_module_handle_private *)(msg_rx->tx_handle),
+					&msg_rx->audio_data);
+			}
+
+			data_fifo_block_free(handle->thread.msg_rx, (void **)(&msg_rx));
+
+			k_mem_slab_free(handle->thread.data_slab, (void **)(&data));
+
+			LOG_ERR("Data process error in module %s, ret %d", handle->name, ret);
+			continue;
+		}
+
+		/* Send processed audio data to next module(s). */
+		send_to_connected_modules(handle, &audio_data);
+
+		if (msg_rx->response_cb != NULL) {
+			msg_rx->response_cb((struct audio_module_handle_private *)msg_rx->tx_handle,
+					    &msg_rx->audio_data);
+		}
+
+		data_fifo_block_free(handle->thread.msg_rx, (void **)&msg_rx);
+	}
+
+	CODE_UNREACHABLE;
+}
+
+/**
+ * @brief  Function for opening a module.
```

Comment:
I still se that there are comments over every non-static function. I dont think this is needed since the functions should be documented in the .h file